File: android/app/src/main/java/be/sportgreenmoove/app/ui/MessagesScreen.kt
package be.sportgreenmoove.app.ui

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.em
import androidx.compose.ui.unit.sp
import be.sportgreenmoove.app.data.InboxSummary
import be.sportgreenmoove.app.design.Sgm
import be.sportgreenmoove.app.design.SgmColor
import be.sportgreenmoove.app.design.SgmType
import be.sportgreenmoove.app.services.FirebaseGateway

@Composable
fun MessagesScreen(firebase: FirebaseGateway) {
    var selectedTab by remember { mutableStateOf("chats") }
    var inbox by remember { mutableStateOf(InboxSummary()) }
    var loading by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }

    LaunchedEffect(firebase) {
        loading = true
        error = null
        runCatching { firebase.getInbox() }
            .onSuccess { inbox = it }
            .onFailure { error = it.message ?: "Inbox indisponible." }
        loading = false
    }

    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Sgm.colors.bgApp)
            .verticalScroll(rememberScrollState())
            .padding(bottom = 20.dp),
    ) {
        MessagesHeader()
        MessagesTabs(tabs = messageTabs(inbox), selected = selectedTab, onSelected = { selectedTab = it })
        Column(
            modifier = Modifier.padding(horizontal = 20.dp),
            verticalArrangement = Arrangement.spacedBy(10.dp),
        ) {
            when {
                loading -> MessagesEmptyCard("Synchronisation de l'inbox...")
                error != null -> MessagesEmptyCard(error.orEmpty())
                selectedTab == "notifs" -> if (inbox.notifications.isEmpty()) {
                        MessagesEmptyCard("Aucune notification.")
                    } else {
                        inbox.notifications.forEach { notice -> NotificationCard(notice) }
                    }
                selectedTab == "avis" -> if (inbox.reviews.isEmpty()) {
                        MessagesEmptyCard("Aucun avis en attente.")
                    } else {
                        inbox.reviews.forEach { review -> ReviewCard(review) }
                    }
                else -> if (inbox.chats.isEmpty()) {
                        MessagesEmptyCard("Aucune conversation.")
                    } else {
                        inbox.chats.forEach { chat -> ChatCard(chat) }
                }
            }
        }
    }
}

@Composable
private fun MessagesHeader() {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(start = 20.dp, top = 34.dp, end = 20.dp, bottom = 16.dp),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(10.dp),
    ) {
        Text(
            "MESSAGES",
            style = SgmType.DisplayXL.copy(color = Sgm.colors.textPrimary, fontSize = 22.sp, letterSpacing = 0.08.em),
            modifier = Modifier.weight(1f),
        )
        V2ThemeButton()
    }
}

@Composable
private fun MessagesTabs(tabs: List<MessageTab>, selected: String, onSelected: (String) -> Unit) {
    Row(
        modifier = Modifier.padding(start = 20.dp, end = 20.dp, bottom = 14.dp),
        horizontalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        tabs.forEach { tab ->
            MessageTabButton(
                tab = tab,
                selected = selected == tab.id,
                onClick = { onSelected(tab.id) },
                modifier = Modifier.weight(1f),
            )
        }
    }
}

@Composable
private fun MessageTabButton(tab: MessageTab, selected: Boolean, onClick: () -> Unit, modifier: Modifier) {
    Row(
        modifier = modifier
            .height(34.dp)
            .clip(RoundedCornerShape(999.dp))
            .background(if (selected) SgmColor.Green else Sgm.colors.bgCard)
            .border(BorderStroke(1.dp, Sgm.colors.border), RoundedCornerShape(999.dp))
            .clickable(onClick = onClick),
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.Center,
    ) {
        Text(
            tab.label,
            style = SgmType.Label.copy(
                color = if (selected) SgmColor.TextOnGreen else Sgm.colors.textMuted,
                fontSize = 10.sp,
                letterSpacing = 0.06.em,
            ),
        )
        if (tab.count > 0) {
            Spacer(Modifier.size(4.dp))
            MessageTabBadge(tab.count.toString(), selected)
        }
    }
}

private fun messageTabs(inbox: InboxSummary) = listOf(
    MessageTab("notifs", "NOTIFS", inbox.notifications.count { it.unread }),
    MessageTab("chats", "CHATS", inbox.chats.sumOf { it.unreadCount }),
    MessageTab("avis", "AVIS", inbox.reviews.size),
)

private data class MessageTab(val id: String, val label: String, val count: Int)

File: ios/Sources/SportsGreenMooveApp/MessagesScreen.swift
import SwiftUI

struct MessagesScreen: View {
    @Environment(AppState.self) private var appState
    @State private var selectedTab = "chats"
    @State private var inbox = InboxSummary()
    @State private var loading = false
    @State private var error: String?

    var body: some View {
        SGMScreen(spacing: 0) {
            SGMTopBar(title: "MESSAGES")
            MessageTabs(tabs: tabs, selected: selectedTab) { selectedTab = $0 }
            VStack(spacing: 10) {
                if loading {
                    MessageEmptyCard(message: "Synchronisation de l'inbox...")
                } else if let error {
                    MessageEmptyCard(message: error)
                } else {
                    switch selectedTab {
                    case "notifs":
                        notificationContent
                    case "avis":
                        reviewContent
                    default:
                        chatContent
                    }
                }
            }
            .padding(.horizontal, SGMSpace.padScreen)
            .padding(.top, 14)
        }
        .task { await loadInbox() }
    }

    private var tabs: [MessageTab] {
        [
            MessageTab(id: "notifs", label: "Notifs", count: inbox.notifications.filter(\.unread).count),
            MessageTab(id: "chats", label: "Chats", count: inbox.chats.reduce(0) { $0 + $1.unreadCount }),
            MessageTab(id: "avis", label: "Avis", count: inbox.reviews.count),
        ]
    }

    private var chatContent: some View {
        Group {
            if inbox.chats.isEmpty {
                MessageEmptyCard(message: "Aucune conversation.")
            } else {
                ForEach(inbox.chats) { ChatCard(chat: $0) }
            }
        }
    }

    private var notificationContent: some View {
        Group {
            if inbox.notifications.isEmpty {
                MessageEmptyCard(message: "Aucune notification.")
            } else {
                ForEach(inbox.notifications) { NotificationCard(notification: $0) }
            }
        }
    }

    private var reviewContent: some View {
        Group {
            if inbox.reviews.isEmpty {
                MessageEmptyCard(message: "Aucun avis en attente.")
            } else {
                ForEach(inbox.reviews) { ReviewCard(review: $0) }
            }
        }
    }

    private func loadInbox() async {
        loading = true
        error = nil
        defer { loading = false }
        do {
            inbox = try await appState.firebase.getInbox()
        } catch {
            self.error = error.localizedDescription
        }
    }
}

private struct MessageTab: Identifiable, Hashable {
    let id: String
    let label: String
    let count: Int
}

private struct MessageTabs: View {
    let tabs: [MessageTab]
    let selected: String
    var onSelect: (String) -> Void

    var body: some View {
        HStack(spacing: 6) {
            ForEach(tabs) { tab in
                SGMChip(text: tab.label, selected: selected == tab.id, badge: tab.count > 0 ? "\(tab.count)" : nil) {
                    onSelect(tab.id)
                }
            }
        }
        .padding(.horizontal, SGMSpace.padScreen)
        .padding(.top, 12)
    }
}

private struct ChatCard: View {
    let chat: InboxChatSummary

    var body: some View {
        HStack(spacing: 10) {
            SGMAvatar(initials: chat.initials, size: 40)
            VStack(alignment: .leading, spacing: 2) {
                Text(chat.title)
                    .font(.sgmBody(15, weight: .semibold))
                    .foregroundStyle(SGM.textPrimary)
                    .lineLimit(1)
                Text(chat.preview)
                    .font(.sgmBody(12))
                    .foregroundStyle(SGM.textMuted)
                    .lineLimit(1)
            }
            Spacer()
            VStack(alignment: .trailing, spacing: 4) {
                Text(chat.dateLabel)
                    .font(.sgmBody(10, weight: .medium))
                    .foregroundStyle(SGM.textMuted)
                if chat.unreadCount > 0 {
                    Text("\(chat.unreadCount)")
                        .font(.sgmBody(10, weight: .bold))
                        .foregroundStyle(SGM.textOnGreen)
                        .frame(width: 18, height: 18)
                        .background(SGM.green, in: Circle())
                }
            }
        }
        .messageCardStyle()
    }
}

private struct NotificationCard: View {
    let notification: InboxNotificationSummary

    var body: some View {
        HStack(alignment: .top, spacing: 10) {
            SGMAvatar(initials: notification.initials, size: 36, muted: !notification.unread)
            VStack(alignment: .leading, spacing: 3) {
                Text(notification.title)
                    .font(.sgmBody(12, weight: .bold))
                    .tracking(.sgmWide(for: 12))
                    .foregroundStyle(notification.unread ? SGM.green : SGM.textMuted)
                Text(notification.body)
                    .font(.sgmBody(13, weight: notification.unread ? .semibold : .regular))
                    .foregroundStyle(SGM.textPrimary)
                Text(notification.dateLabel)
                    .font(.sgmBody(10, weight: .medium))
                    .foregroundStyle(SGM.textMuted)
            }
            Spacer()
            if notification.unread {
                Circle().fill(SGM.green).frame(width: 8, height: 8).padding(.top, 4)
            }
        }
        .messageCardStyle()
    }
}

private struct ReviewCard: View {
    let review: InboxReviewPrompt

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack(spacing: 10) {
                SGMAvatar(initials: review.initials, size: 36)
                Text(review.title)
                    .font(.sgmBody(13, weight: .bold))
                    .foregroundStyle(SGM.textPrimary)
            }
            Text(review.prompt)
                .font(.sgmBody(13))
                .foregroundStyle(SGM.textSecondary)
            HStack(spacing: 4) {
                ForEach(0..<5, id: \.self) { _ in
                    SGMIconView(icon: .star, size: 18, color: SGM.orange)
                }
            }
        }
        .messageCardStyle()
    }
}

private struct MessageEmptyCard: View {
    let message: String

    var body: some View {
        Text(message)
            .font(.sgmBody(13, weight: .semibold))
            .foregroundStyle(SGM.textMuted)
            .frame(maxWidth: .infinity, alignment: .leading)
            .messageCardStyle()
    }
}

private extension View {
    func messageCardStyle() -> some View {
        padding(.horizontal, 14)
            .padding(.vertical, 12)
            .background(SGM.bgCard, in: RoundedRectangle(cornerRadius: 14, style: .continuous))
            .overlay(RoundedRectangle(cornerRadius: 14, style: .continuous).stroke(SGM.border, lineWidth: 1))
    }
}
